src/utils/Utils.ts

```diff
@@ -194,6 +265,118 @@ export default class Utils {
   public static translateToNaturalLanguage(text: string): string {
     return (translations as any)[text];
   }
+
+  public static async recipeAttributesToList(recipe: Recipe): Promise<ListEntry[]> {
+    return [
+      {
+        id: 'name',
+        firstValue: 'Name',
+        secondValue: recipe.name,
+      },
+      {
+        id: 'Input Ingrediants',
+        firstValue: 'Input Ingrediants',
+        secondValue: recipe.ingredients ? await this.getInputIngrediantList(recipe.ingredients) : '',
+      },
+      {
+        id: 'transformerIngredients',
+        firstValue: 'Transformer Ingrediants',
+        secondValue: recipe.ingredients ? await this.getTransformerIngrediantList(recipe.ingredients) : '',
+      },
+      {
+        id: 'algorithmIngredient',
+        firstValue: 'Algorithm Ingrediants',
+        secondValue: recipe.ingredients ? await this.getAlgorithmIngrediantList(recipe.ingredients) : '',
+      },
+      {
+        id: 'outputIngredients',
+        firstValue: 'Output Ingrediants',
+        secondValue: recipe.ingredients ? await this.getOutputIngrediantList(recipe.ingredients) : '',
+      },
+    ];
+  }
+
+  public static async getInputIngrediantList(ingredients: Ingredient[]) {
+    const inputIngredients = ingredients.filter( (ingredient) => {
+      return ingredient.ingredientType === IngredientType.INPUT;
+    });
+    let inputList = '';
+    const getNames = async () => {
```

Comment:
There is a nice construct for that called Promise.all( iterable: Promise[] ). Here is the documentation for that: https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Promise/all.

This is what you want to do here:
```ts
const namePromises = inputIngredients.map(async (ingredient: Ingredient): Promise<string> => {
  const resourceType = await ResourceTypes.getOne(ingredient.ingredientDefinition!);
  return resourceType.name;
});
const names = await Promise.all(namePromises);
return names.join(', ');
```

This way your code is not only shorter, but it also executes in a true-async behavior (without waiting in for loops). Please adjust the other util functions below accordingly to this example.

Reply:
I really liked the asyncForEach, but this is better.

Reply:
thanks for pointing this out!